### PR TITLE
Reporting updates

### PR DESCRIPTION
**Pull request by andrewjl-mux**
Adds events for input standardization success and failure and upload failure using a revised event schema. Upload success also updated to match the revised schema.

Request headers in separate PR https://github.com/muxinc/swift-upload-sdk/pull/67.

This PR also includes a workaround for an issue with [`AVAssetExportSession`](https://developer.apple.com/documentation/avfoundation/avassetexportsession) where the asynchronously loading the duration from the standardized file hangs / takes a very long time. As part of the workaround, the duration of the original input file rather than the standardized input file is used when reporting.

### CODE REVIEW

File: Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift [hunk 1]
```diff
@@ -33,83 +40,248 @@ class Reporter: NSObject {
         let jsonEncoder = JSONEncoder()
         jsonEncoder.keyEncodingStrategy = JSONEncoder.KeyEncodingStrategy.convertToSnakeCase
         jsonEncoder.outputFormatting = .sortedKeys
+        jsonEncoder.dateEncodingStrategy = .iso8601
```
Comment (andrewjl-mux) [inline Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:43]: Dates are automatically formatted to be like this: `2023-07-07T04:12:48Z`

File: Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift [hunk 2]
```diff
@@ -33,83 +40,248 @@ class Reporter: NSObject {
...
 extension Reporter: URLSessionDelegate, URLSessionTaskDelegate {
     func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Swift.Void) {
-        if(self.pendingUploadEvent != nil) {
-            if let redirectUrl = request.url, let httpBody = try? serializePendingEvent() {
-                let request = self.generateRequest(
-                    url: redirectUrl,
-                    httpBody: httpBody
-                )
-                completionHandler(request)
+        if let pendingEvent = pendingEvents[ObjectIdentifier(task)], let redirectURL = request.url {
```
Comment (andrewjl-mux) [inline Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:264]: `Reporter` can now manage multiple in-flight requests and can now be shared, this is a very simplistic implementation, not thread-safe, but will work reasonably outside of weird edge cases.

File: Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift [hunk 3]
```diff
@@ -33,83 +40,248 @@ class Reporter: NSObject {
...
+                httpBody: httpBody
+            )
+
+            completionHandler(request as URLRequest)
         }
     }
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        pendingEvents[
+            ObjectIdentifier(task)
```
Comment (andrewjl-mux) [inline Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:284]: We rely on a hash of the `URLSessionTask` pointer ([`ObjectIdentifier`](https://developer.apple.com/documentation/swift/objectidentifier)) to correlate a request with a pending event.

File: Sources/MuxUploadSDK/Upload/ChunkedFileUploader.swift [hunk 4]
```diff
@@ -131,7 +133,43 @@ class ChunkedFileUploader {
...
+                        lastUpdate = update
+                    } else {
+                        lastUpdate = nil
+                    }
+
+                    // This modifies currentState, so capture
+                    // the last update first
                     notifyStateFromWorker(.failure(uploadError))
+
+                    // FIXME: Will only work if currentState
```
Comment (andrewjl-mux) [inline Sources/MuxUploadSDK/Upload/ChunkedFileUploader.swift:148]: Workaround done in order to minimize changing `ChunkedFileUploader` internals, suggestions for improvement very welcome

File: Sources/MuxUploadSDK/PublicAPI/MuxUpload.swift [hunk 5]
```diff
@@ -475,6 +481,19 @@ public final class MuxUpload : Hashable, Equatable {
         if !uploadInfo.options.inputStandardization.isEnabled {
             startNetworkTransport(videoFile: videoFile)
         } else {
+            let inputStandardizationStartTime = Date()
+            let reporter = Reporter.shared
+
+            // For consistency report non-std
```
Comment (andrewjl-mux) [inline Sources/MuxUploadSDK/PublicAPI/MuxUpload.swift:487]: Retrospectively, now think we should report the standardized input size if we can. I'll update this PR with that change.

File: Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift [hunk 6]
```diff
@@ -33,83 +40,244 @@ class Reporter: NSObject {
...
+            let request = NSMutableURLRequest.makeJSONPost(
+                url: redirectURL,
+                httpBody: httpBody
+            )
+
+            completionHandler(request as URLRequest)
         }
     }
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
```
Comment (daytime-em) [inline Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:278]: Do we need to clear the pending task when the event was sent successfully also?

**Review by daytime-em — CHANGES_REQUESTED** [review]
I just had the one concern, otherwise it looks great!!

File: Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift (hunk 6 above)
Comment (andrewjl-mux) [reply Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:278]: You're right that we do and that's already handled in this method:

```
pendingEvents[
            ObjectIdentifier(task)
] = nil
```

Comment (andrewjl-mux) [reply Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:278]: `URLSession` [always calls this delegate method](https://developer.apple.com/documentation/foundation/urlsessiontaskdelegate/1411610-urlsession), when the task completes successfully the error parameter is nil, so the logic there now covers that case.

Comment (daytime-em) [reply Sources/MuxUploadSDK/InternalUtilities/Reporting/Reporter.swift:278]: Ahh, now I understand. Thanks for the explanation!

**Comment by andrewjl-mux** [comment]
@daytime-em I think you meant to approve this instead of merging, right? Checking to make sure all the changes requested were actually done.

**Comment by daytime-em** [comment]
Aa!! Sorry, I did want to approve it. Oops!

**Pull request merged**